test/unit/wireclient.test.mjs

```diff
@@ -0,0 +1,720 @@
+      connectedClient.configure({ storageApi: 'draft-dejong-remotestorage-02' });
+      await connectedClient.put('/foo/bar', '1', 'text/plain', { ifNoneMatch: 'etag1' });
+      const call = fetchMock.calls('putFileOK')[0];
+      expect(call[1].headers['If-None-Match']).to.equal('"etag1"');
+      expect(call[1].headers).not.to.have.property('If-Match');
+    });
+
+    it('sends If-Match header when revisions are supported and rev is given', async () => {
+      connectedClient.configure({ storageApi: 'draft-dejong-remotestorage-02' });
+      await connectedClient.put('/foo/bar', '1', 'text/plain', { ifMatch: 'etag2' });
+      const call = fetchMock.calls('putFileOK')[0];
+      expect(call[1].headers['If-Match']).to.equal('"etag2"');
+      expect(call[1].headers).not.to.have.property('If-None-Match');
+    });
+
+    it('encodes special characters in the path', async () => {
+      const url = `${BASE_URI}/foo/A%252FB/bar`;
+      fetchMock.mock(
+        { name: 'putEncoded', url, method: 'PUT' },
+        { status: 200 }
+      );
+
+      await connectedClient.put('/foo/A%2FB/bar', 'baz', 'text/plain');
+      const calls = fetchMock.calls('putEncoded');
+      expect(calls).to.have.lengthOf(1);
+      expect(calls[0][0]).to.equal(url);
+    });
+
+    it('encodes spaces in the path', async () => {
+      const url = `${BASE_URI}/foo/A%20B/bar`;
+      fetchMock.mock(
+        { name: 'putWithSpace', url, method: 'PUT' },
+        { status: 200 }
+      );
+
+      await connectedClient.put('/foo/A B/bar', 'baz', 'text/plain');
+      const calls = fetchMock.calls('putWithSpace');
+      expect(calls).to.have.lengthOf(1);
+      expect(calls[0][0]).to.equal(url);
+    });
+
+    it('leaves slash characters in the path', async () => {
+      const url = `${BASE_URI}/test/foo/A/B/C/D/E`;
+      fetchMock.mock(
+        { name: 'putWithSlashes', url, method: 'PUT' },
+        { status: 200 }
+      );
+
+      await connectedClient.put('/test/foo/A/B/C/D/E', 'slash it!', 'text/plain');
+      const calls = fetchMock.calls('putWithSlashes');
+      expect(calls).to.have.lengthOf(1);
+      expect(calls[0][0]).to.equal(url);
+    });
+
+    it('removes redundant slash characters in the path', async () => {
+      const url = `${BASE_URI}/test/foo/A/B/C/D/E`;
+      fetchMock.mock(
+        { name: 'putWithSlashes', url, method: 'PUT' },
+        { status: 200 }
+      );
+
+      await connectedClient.put('/test/foo/A/B/C///D/E', 'slash it!', 'text/plain');
+      const calls = fetchMock.calls('putWithSlashes');
+      expect(calls).to.have.lengthOf(1);
+      expect(calls[0][0]).to.equal(url);
+    });
+
+    it('adds no charset for strings', async () => {
+      await connectedClient.put('/foo/bar', 'something', 'text/html');
+      const call = fetchMock.calls('putFileOK')[0];
+      expect(call[1].headers).to.have.property('Content-Type').which.equals('text/html');
+    });
+
+    it('preserves charset for strings', async () => {
```

Comment:
Do we need a separate test for "PUTs of strings have custom charset preserved" ?